File: bab_11/log.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <sys/stat.h>
#include <string.h>
#include <errno.h>
#include <stdarg.h>

#include "log.h"

void create_log_directory() {
    struct stat st;
    if (stat(LOG_DIR, &st) != 0) { // Cek apakah folder log sudah ada
        if (mkdir(LOG_DIR, 0700) != 0) { // Buat folder log jika belum ada
            fprintf(stderr, "Warning: Failed to create log directory: %s\n", strerror(errno));
            // Tidak keluar program, biarkan log tetap berjalan atau lakukan penanganan lebih lanjut
        }
    }
}

void write_log(const char *format, ...) {
    time_t now = time(NULL);
    struct tm *t = localtime(&now);

    // Format tanggal untuk nama file log
    char log_filename[100];
    snprintf(log_filename, sizeof(log_filename), LOG_DIR "%04d-%02d-%02d.log",
             t->tm_year + 1900, t->tm_mon + 1, t->tm_mday); // Simpan di dalam folder logs

    FILE *log_file = fopen(log_filename, "a"); // Buka file log untuk menambahkan
    if (log_file != NULL) {
        // Format timestamp dalam format yang sesuai
        char timestamp[25];
        strftime(timestamp, sizeof(timestamp), "%Y-%m-%d %H:%M:%S", t);
        
        // Tulis timestamp ke file log
        fprintf(log_file, "[%s] ", timestamp);

        // Mengelola variadic arguments
        va_list args;
        va_start(args, format);
        vfprintf(log_file, format, args); // Tulis pesan dengan format variadic ke file log
        va_end(args);

        fprintf(log_file, "\n"); // Tambahkan newline otomatis
        fclose(log_file); // Tutup file
    } else {
        fprintf(stderr, "Failed to open log file: %s\n", strerror(errno));
    }
}

File: bab_9/python/log.py
import os
# import time
from datetime import datetime
from config import cfg

def create_log_directory():
    """Cek apakah folder log sudah ada, jika belum buat folder tersebut."""
    if not os.path.exists(cfg.log_directory):
        try:
            os.mkdir(cfg.log_directory)
        except OSError as e:
            print(f"Warning: Failed to create log directory: {e}")

def write_log(format_string, *args):
    """Tulis pesan log ke file log dengan format timestamp."""
    # Format tanggal untuk nama file log
    log_filename = os.path.join(cfg.log_directory, f"{datetime.now():%Y-%m-%d}.log")

    try:
        with open(log_filename, "a") as log_file:
            # Format timestamp
            timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            # Tulis timestamp dan pesan log ke file
            log_file.write(f"[{timestamp}] {format_string % args}\n")
    except OSError as e:
        print(f"Failed to open log file: {e}")

# Contoh penggunaan
# create_log_directory()
# write_log("Log message: %s", "Program started")
# write_log("Another log message: %d", 42)
